### DISCUSSION TITLE
local variable 'pyarmor_core_445' referenced before assignment

### DISCUSSION THREAD

**Discussion by xuyingliang** [dashingsoft/pyarmor#1565] [Q&A] [answered]
使用pyarmor加密项目，出现如下错误，请问是什么原因。
pyarmor -d gen -r -i --enable-jit --assert-call --platform windows.x86_64 --platform linux.x86_64 QC

pyarmor.error.log内容如下：

10556 MainProcess 2023-11-01 16:01:26,498
Traceback (most recent call last):
  File "C:\Users\xyl\AppData\Local\Programs\Python\Python38\lib\site-packages\pyarmor\cli\__main__.py", line 712, in main
    main_entry(sys.argv[1:])
  File "C:\Users\xyl\AppData\Local\Programs\Python\Python38\lib\site-packages\pyarmor\cli\__main__.py", line 700, in main_entry
    return args.func(ctx, args)
  File "C:\Users\xyl\AppData\Local\Programs\Python\Python38\lib\site-packages\pyarmor\cli\__main__.py", line 230, in cmd_gen
    builder.process(options)
  File "C:\Users\xyl\AppData\Local\Programs\Python\Python38\lib\site-packages\pyarmor\cli\generate.py", line 147, in process
    Pytransform3.pre_build(self.ctx)
  File "C:\Users\xyl\AppData\Local\Programs\Python\Python38\lib\site-packages\pyarmor\cli\core\__init__.py", line 118, in pre_build
    return m.pre_build(ctx)
  File "<maker>", line 681, in pre_build
  File "<maker>", line 621, in build
  File "<maker>", line 3199, in process
  File "<maker>", line 3134, in _get_export_names
UnboundLocalError: local variable 'pyarmor_core_445' referenced before assignment

**Comment by jondy** [comment, marked as answer]
使用下面的命令禁用局部变量的混淆
```
pyarmor cfg mix_localnames = 0
```
然后在进行加密试试

**Reply by xuyingliang** [reply]
使用后，还是出这个错误提示，请问是什么问题？

F:\PublicForm>pyarmor cfg mix_localnames = 0
INFO     Python 3.8.3
INFO     Pyarmor 8.4.1 (group), 006064, 教学实验平台
INFO     Platform windows.x86_64
INFO     change option "mix_localnames" to new value "0"
------------------------------------------------------------
Section: builder
Current settings
  mix_localnames = 1
Global settings
Local settings
  mix_localnames = 0
F:\PublicForm>pyarmor -d gen -r -i --enable-jit --assert-call --platform windows.x86_64 --platform linux.x86_64 QcFrame
INFO     Python 3.8.3
INFO     Pyarmor 8.4.1 (group), 006064, 教学实验平台
INFO     Platform windows.x86_64
DEBUG    native platform windows.amd64
DEBUG    home path: C:\Users\xyl\.pyarmor
INFO     use runtime platforms: ['windows.x86_64', 'linux.x86_64']
DEBUG    implicitly set restrict_module = 2
DEBUG    command options: {'recursive': True, 'no_runtime': False, 'enable_jit': True, 'assert_call': True, 'import_prefix': 1, 'platforms': ['windows.x86_64', 'linux.x86_64'], 'inputs': ['QcFrame'], 'restrict_module': 2}
DEBUG    install plugin: CodesignPlugin
DEBUG    install plugin: DarwinUniversalPlugin
INFO     search inputs ...
INFO     find package at QcFrame
INFO     find 1 top resources
DEBUG    build package relations
DEBUG    unknown error, please check pyarmor.error.log
ERROR    local variable 'pyarmor_core_445' referenced before assignment

**Comment by jondy** [comment]
把 `--assert-call` 选项去了吧

**Reply by xuyingliang** [reply]
去掉--assert-call就不出上面那个错了。
在加密时，遇到第三方包出“GBK”的问题，请问怎样跳过去。
QcFrame\main
               \utils
              \models
              \libs

我不想加密libs这个目录，应该怎样处理。

pyarmor -d gen -r -i --enable-jit --platform windows.x86_64 --platform linux.x86_64 QcFrame

**Comment by jondy** [comment]
自己看看文档吧

**Reply by xuyingliang** [reply]
执行：pyarmor cfg QcFrame.libs ofb_code=0
pyarmor -d gen -r -i --enable-jit --platform windows.x86_64 --platform linux.x86_64 QcFrame
还是会找到libs目录下的文件，Error ’gbk' codec can't decode byte 0x84 in position 3756: illegal multibyte sequence


想采用公共的运行辅助包，单个目录进行加密，又出ERROR    --outer is required if using --no_runtime：
F:\PublicForm>pyarmor gen runtime -O build_product_runtime --platform windows.x86_64 --platform linux.x86_64
INFO     Python 3.8.3
INFO     Pyarmor 8.4.1 (group), 006064, 教学实验平台
INFO     Platform windows.x86_64
INFO     use runtime platforms: ['windows.x86_64', 'linux.x86_64']
INFO     start to generate runtime package
INFO     target platforms {'linux.x86_64', 'windows.x86_64'}
INFO     write build_product_runtime\pyarmor_runtime_006064\linux_x86_64\pyarmor_runtime.so
INFO     patch runtime file
INFO     write build_product_runtime\pyarmor_runtime_006064\windows_x86_64\pyarmor_runtime.pyd
INFO     patch runtime file
INFO     write "build_product_runtime\.pyarmor.ikey"
INFO     generate runtime package to "build_product_runtime" OK


F:\PublicForm>pyarmor -d gen -r --use-runtime build_product_runtime --enable-jit --platform windows.x86_64 --platform linux.x86_64 QcFrame
INFO     Python 3.8.3
INFO     Pyarmor 8.4.1 (group), 006064, 教学实验平台
INFO     Platform windows.x86_64
DEBUG    native platform windows.amd64
DEBUG    home path: C:\Users\xyl\.pyarmor
INFO     use runtime platforms: ['windows.x86_64', 'linux.x86_64']
DEBUG    command options: {'recursive': True, 'no_runtime': True, 'enable_jit': True, 'platforms': ['windows.x86_64', 'linux.x86_64'], 'inputs': ['QcFrame'], 'use_runtime': 'build_product_runtime'}
ERROR    --outer is required if using --no_runtime

**Comment by jondy** [comment]
这个问题下一个版本会解决的

**Reply by xuyingliang** [reply]
谢谢！
